resources/bundles/org.eclipse.core.filesystem/src/org/eclipse/core/internal/filesystem/local/LocalFile.java

```diff
@@ -129,26 +131,45 @@ public String[] childNames(int options, IProgressMonitor monitor) {
 
 	@Override
 	public void copy(IFileStore destFile, int options, IProgressMonitor monitor) throws CoreException {
-		if (destFile instanceof LocalFile) {
-			File source = file;
-			File destination = ((LocalFile) destFile).file;
+		if (destFile instanceof LocalFile destination) {
 			//handle case variants on a case-insensitive OS, or copying between
 			//two equivalent files in an environment that supports symbolic links.
 			//in these nothing needs to be copied (and doing so would likely lose data)
 			try {
-				if (isSameFile(source, destination)) {
+				if (isSameFile(this.file, destination.file)) {
 					//nothing to do
 					return;
 				}
 			} catch (IOException e) {
-				String message = NLS.bind(Messages.couldNotRead, source.getAbsolutePath());
+				String message = NLS.bind(Messages.couldNotRead, this.file.getAbsolutePath());
 				Policy.error(EFS.ERROR_READ, message, e);
 			}
 		}
 		//fall through to super implementation
 		super.copy(destFile, options, monitor);
 	}
 
+	private static final CopyOption[] NO_OVERWRITE = {StandardCopyOption.COPY_ATTRIBUTES};
+	private static final CopyOption[] OVERWRITE_EXISTING = {StandardCopyOption.COPY_ATTRIBUTES, StandardCopyOption.REPLACE_EXISTING};
+
+	@Override
+	protected void copyFile(IFileInfo sourceInfo, IFileStore destination, int options, IProgressMonitor monitor) throws CoreException {
+		if (destination instanceof LocalFile target) {
+			try {
+				boolean overwrite = (options & EFS.OVERWRITE) != 0;
+				Files.copy(this.file.toPath(), target.file.toPath(), overwrite ? OVERWRITE_EXISTING : NO_OVERWRITE);
+			} catch (FileAlreadyExistsException e) {
+				Policy.error(EFS.ERROR_EXISTS, NLS.bind(Messages.fileExists, target.filePath, e));
```

Comment:
I think this is a misplaced parenthesis. If I am not mistaken, the exception has to passed to `Policy.error()` instead of the NLS binding. The NLS does not have a second parameter and the `error()` method accepts the exception as an additional parameter.
```suggestion
				Policy.error(EFS.ERROR_EXISTS, NLS.bind(Messages.fileExists, target.filePath), e);
```

Reply:
That absolutely right. Thanks for spotting this! Just fixed it.